primitives/src/adapter.rs

```diff
@@ -56,11 +52,6 @@ pub struct Session {
 }
 
 pub trait Adapter: ChannelValidator + Clone + Debug {
-    type Output;
```

Comment:
The reason for this not being very good idea is that you can't use something like:

let adapter: Box<dyn Adapter> = match predicate {
 .... => one adapter instantiation
 .... => another adapter instantiation
}

and this is because you need to specify an Output, i.e. Box<dyn Adapter<Output=????>> and you can't write any output, because you have different ones in the `match` arms.

Reply:
I don't think this is unique to this implementation alone

Reply:
You can find more about the associated type error here: https://doc.rust-lang.org/nightly/error-index.html#E0191
if you want to specify `Box<dyn Adapter>` you need to specify the associated type as well.